Add support for version 2 RVFI-DII trace format

arichardson opened the pull request:
See https://github.com/CTSRD-CHERI/QuickCheckVEngine/pull/7 and https://github.com/rems-project/sail-riscv/pull/76

arichardson committed:
Add support for reporting version 2 RVFI-DII traces

This trace format is more extensible and will allow reporting values for
CHERI registers, FP registers, CSRs, additional metadata for traps, etc.

arichardson committed:
Fix missing reporting of CHERI register updates

We also need to update available_fields to indicate that the next trace
packet should contain the integer metadata. This was previously only done
for the values updated by TCG code.

arichardson committed:
Move helper_log_value from MIPS to target independent code

This can be rather useful when debugging TCG code generation issues.

arichardson committed:
Use stores of the right size when updating RVFI DII state

I was seeing some fields being reported as zeroes even though they should
have been set. Setting a hardware watchpoint showed me that they were being
written again from TCG-generated codewith different values after writing
the initial (correct) value. It turns out they were being overwritten by
subsequent stores to other fields since we were unconditionally using a
64-bit store to a given offset.

This previously did not matter since all fields were i64, but now we have
i8, i32 and i64 fields, so we need overloads. Also add some static_asserts
using __builtin_types_compatible_p() to check that we store to the right
field.

arichardson committed:
RVFI-DII: report IXL

arichardson committed:
Improve RVFI-DII debug output

When instruction tracing is enabled, we flush the instruction log inside
rvfi_dii_communicate, so that the side-effects are printed before sending
the packet rather than when we return to the TCG main loop. This also adds
a call to qemu_log_instr_drop() since otherwise the call to
qemu_log_instr_commit() in the translator loop would print another invalid
instruction.

arichardson committed:
RVFI-DII: print whether the instruction was a trap in the debug output

Also pass this information as an argument to the rvfi_dii_communicate
function.

arichardson committed:
RVFI-DII: fix memory data packet layout

The address is the last field and not the first one.

arichardson commented on target/riscv/cpu.c:
```diff
@@ -364,43 +364,139 @@ void restore_state_to_opc(CPURISCVState *env, TranslationBlock *tb,
+        // Print the instruction now and skip the next commit() call that
+        // happens when we return to the translator loop.
+        qemu_log_instr_commit(env);
+        qemu_log_instr_drop(env); // Avoid an invalid instruction log
```
@qwattash is this expected to work? The output looks sane to me and the next call to commit in the translator should clear the force_drop flag.

qwattash replied:
I'm not familiar with the single stepping and RVFI_DII, how does this interact with the normal translator loop, are we injecting multiple instructions before we return to the translator loop?
I think this solution should work, but I think it will waste 1 slot in the ring buffer every time, if tracing in buffered mode.

arichardson replied:
We inject one instruction, after that executes (but before the commit helper is called) it raises a singlestep trap.
That then calls this function. I added this here because I want the instruction changes to be logged before the rvfi packet data is sent so that I can compare what is being sent to what the log says.

qwattash replied:
I see, so after the rvfi code returns we will go back to the loop and the normal `commit()` will be called. Using the force_drop flag here is fine. After checking, I don't think this will cause disruption to the ring buffer as the force_drop flag will cause the next commit to clear the buffer entry.

qwattash reviewed (approved):
Logging looks fine to me

arichardson committed:
RVFI-DII: Fix reporting of capability memory loads/stores

We weren't setting the RVFI_MEM_DATA flag so the memory load/store data
was not being reported.
